File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void generate(vector<int>& els, int t, int i, vector<vector<int>>& ans, vector<int> arr) {
        if(t == 0) {ans.push_back(arr); return;}
        if(i == els.size() || t<0) return;
        arr.push_back(els[i]);
        generate(els, t-els[i], i, ans, arr);
        arr.pop_back();
        generate(els, t, i+1, ans, arr);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> arr;
        generate(candidates, target, 0, ans, arr);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    void ps(int[] arr, int i, List<Integer> ans, List<List<Integer>> allSets, int t) {
        if(i == arr.length || t<0) {
            return;
        }
        if(t == 0) {
            allSets.add(new ArrayList<>(ans));
            return;
        }
        ans.add(arr[i]);
        ps(arr, i, ans, allSets, t-arr[i]);
        ans.remove(ans.size()-1);
        
        ps(arr, i+1, ans, allSets, t);
    }
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> allSets = new ArrayList<>();
        List<Integer> ans = new ArrayList<>();
        ps(nums, 0, ans, allSets, target);
        return allSets;
    }
}